### PR TITLE
contrib/bcc: new package (0.30.0)

### PR DESCRIPTION
**Pull request by nekopsykose**

### CODE REVIEW

File: contrib/bcc/patches/doc-loc.patch [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ add_custom_target(man ALL DEPENDS ${GZFILES})
+-install(FILES ${GZFILES} DESTINATION share/bcc/man/man8)
++install(FILES ${GZFILES} DESTINATION share/man/man8)
+--- a/tools/CMakeLists.txt
++++ b/tools/CMakeLists.txt
+@@ -21,6 +21,6 @@
+   endif()
+ endforeach()
+ install(FILES ${C_FILES} DESTINATION share/bcc/tools)
+-install(FILES ${TXT_FILES} DESTINATION share/bcc/tools/doc)
```
Comment (nekopsykose) [inline contrib/bcc/patches/doc-loc.patch:15]: /usr/share/bcc/tools has these:  

```
argdist*
bashreadline*
bindsnoop*
biolatency*
biolatpcts*
biopattern*
biosnoop*
biotop*
bitesize*
bpflist*
btrfsdist*
btrfsslower*
cachestat*
cachetop*
capable*
cobjnew*
compactsnoop*
cpudist*
cpuunclaimed*
criticalstat*
dbslower*
dbstat*
dcsnoop*
dcstat*
deadlock*
deadlock.c
dirtop*
doc/
drsnoop*
execsnoop*
exitsnoop*
ext4dist*
ext4slower*
f2fsslower*
filegone*
filelife*
fileslower*
filetop*
funccount*
funcinterval*
funclatency*
funcslower*
gethostlatency*
hardirqs*
inject*
javacalls*
javaflow*
javagc*
javaobjnew*
javastat*
javathreads*
killsnoop*
klockstat*
kvmexit*
lib/
llcstat*
mdflush*
memleak*
mountsnoop*
mysqld_qslower*
netqtop*
netqtop.c
nfsdist*
nfsslower*
nodegc*
nodestat*
offcputime*
offwaketime*
old/
oomkill*
opensnoop*
perlcalls*
perlflow*
perlstat*
phpcalls*
phpflow*
phpstat*
pidpersec*
ppchcalls*
profile*
pythoncalls*
pythonflow*
pythongc*
pythonstat*
rdmaucma*
readahead*
reset-trace*
rubycalls*
rubyflow*
rubygc*
rubyobjnew*
rubystat*
runqlat*
runqlen*
runqslower*
shmsnoop*
slabratetop*
sofdsnoop*
softirqs*
solisten*
sslsniff*
stackcount*
statsnoop*
swapin*
syncsnoop*
syscount*
tclcalls*
tclflow*
tclobjnew*
tclstat*
tcpaccept*
tcpcong*
tcpconnect*
tcpconnlat*
tcpdrop*
tcplife*
tcpretrans*
tcprtt*
tcpstates*
tcpsubnet*
tcpsynbl*
tcptop*
tcptracer*
threadsnoop*
tplist*
trace*
ttysnoop*
vfscount*
vfsstat*
virtiostat*
wakeuptime*
wqlat*
xfsdist*
xfsslower*
zfsdist*
zfsslower*
```

the .c seems garbage, but all of these technically could also be in PATH. though i guess nobody else does that either so this is fine

Comment (nekopsykose) [reply contrib/bcc/patches/doc-loc.patch:15]: ah. all of these (python/bash) tools were rewritten in C and are present in libbpf-tools/ subfolder with a separate makefile. some distros package those (with a bpf- prefix). i guess i could do that, maybe in a separate template

**Comment by nekopsykose** [comment]
the libbpf-based libbpf-tools require btf info; i'll test this again after getting some

the python/bash ones from bcc itself work fine


not sure if it would be wanted to split python-bcc and make the parent depend on it

**Comment by q66** [comment]
i enabled the btf info in the kernels earlier (builds fine in 6.6 and 6.8 after i had fixed pahole, haven't tested on linux-steamdeck yet which is older), just need the next kernel patch ver

**Comment by nekopsykose** [comment]
yea, i meant i'm rebuilding it locally to test it, and then this is done

**Comment by nekopsykose** [comment]
> haven't tested on linux-steamdeck yet which is older  

since it used to fail before it might still be broken for the (newer) 6.5 kernel they use now, but i think nobody is going to really use that on steamdeck kernel so no biggie

**Comment by q66** [comment]
i think the failure might have been just pahole being busted, it would crash but the crash didn't prevent the build from continuing which would result in a weird compiler error later (i don't remember if that's what i was encountering, but it's possible)

**Comment by nekopsykose** [comment]
hopefully just that then :D

**Comment by q66** [comment]
i'll give it a build in like an hour

**Comment by nekopsykose** [comment]
ok, reasonably happy with this- now one can use python-bcc directly without pulling in bash/default sample programs. just libbpf-tools testing to go, will test aarch64/ppc64le builds too

**Comment by nekopsykose** [comment]
builds on both (sans luajit for ppc64le), works on aarch64 too  

on ppc64le:  

```

[1/233] Generating bcc.lua
Writing bcc.lua...
OK!
[2/233] Generating bcc.o
/usr/bin/luajit: /usr/share/luajit-2.1/jit/bcsave.lua:241: assertion failed!
```

that is

```
  local ai = assert(map_arch[ctx.arch])
```

which is

```
local map_arch = {
  x86 =		{ e = "le", b = 32, m = 3, p = 0x14c, },
  x64 =		{ e = "le", b = 64, m = 62, p = 0x8664, },
  arm =		{ e = "le", b = 32, m = 40, p = 0x1c0, },
  arm64 =	{ e = "le", b = 64, m = 183, p = 0xaa64, },
  arm64be =	{ e = "be", b = 64, m = 183, },
  ppc =		{ e = "be", b = 32, m = 20, },
  mips =	{ e = "be", b = 32, m = 8, f = 0x50001006, },
  mipsel =	{ e = "le", b = 32, m = 8, f = 0x50001006, },
  mips64 =	{ e = "be", b = 64, m = 8, f = 0x80000007, },
  mips64el =	{ e = "le", b = 64, m = 8, f = 0x80000007, },
  mips64r6 =	{ e = "be", b = 64, m = 8, f = 0xa0000407, },
  mips64r6el =	{ e = "le", b = 64, m = 8, f = 0xa0000407, },
  s390x =	{ e = "be", b = 64, m = 22, },
}
```

so seems there is no ppc impl for bytecode saving. or maybe it would work by expanding the arch. in any case, left it out for ppc

**Comment by nekopsykose** [comment]
never seen anyone use bcc-lua anyway, just added it for completeness

**Comment by q66** [comment]
https://github.com/void-linux/void-packages/blob/master/srcpkgs/LuaJIT/patches/fix-bcsave-ppc64.patch this could work (my old patch, probably never made it into openresty)

**Comment by nekopsykose** [comment]
none of it applies but i will rebase it i guess

**Comment by nekopsykose** [comment]
yep, it was that easy. but i can't test at runtime since i have no root for bpf on ppc.

though it doesn't really work anyway:  

```

$ doas bcc-lua x.lua
[ERROR] bcc.lua:2070: Symbol not found: bpf_module_create_c_from_string
stack traceback:
	[C]: in function '__index'
````

those are symbols from bcc source files themselves, so i'm not sure how they're meant to work. in the .lua libbcc comes in via require() and i guess squishy is meant to do some magic. it's not strip/lto that breaks it..

**Comment by nekopsykose** [comment]
```console
$ doas bpf-biopattern 
libbpf: failed to determine tracepoint 'block/block_rq_complete' perf event ID: No such file or directory
libbpf: prog 'handle__block_rq_complete': failed to create tracepoint 'block/block_rq_complete' perf event: No such file or directory
libbpf: prog 'handle__block_rq_complete': failed to auto-attach: -2
failed to attach BPF programs
$ doas mount -t debugfs none /sys/kernel/debug
$ doas bpf-biopattern
Tracing block device I/O requested seeks... Hit Ctrl-C to end.
DISK     %RND  %SEQ    COUNT     KBYTES
^Csda        95     4       21        264
nvme0n1    98     1     6852     113892
```

and we're in business :)

so just the lua mystery left. but i could also just yeet the whole thing since it doesn't matter (and leave the bcsave patch since it seems to work)

**Comment by nekopsykose** [comment]
```console
$ doas bpf-cpufreq 
Sampling CPU freq system-wide & by process. Ctrl-C to end.
^C
     kworker/u64:7 : count     distribution
        3800       : 1        |****************************************|

     pw-data-loop  : count     distribution
        3400       : 3        |************************                |
        3800       : 5        |****************************************|
        4400       : 1        |********                                |

     eza           : count     distribution
        4400       : 1        |****************************************|

     swaync-client : count     distribution
        3800       : 1        |****************************************|

     kworker/u65:1 : count     distribution
        3800       : 2        |****************************************|

     dino          : count     distribution
        400        : 1        |****************************************|

     WebExtensions : count     distribution
        400        : 1        |****************************************|

     htop          : count     distribution
        3800       : 6        |****************************************|

     Isolated Web Co : count     distribution
        400        : 5        |*********************************       |
        3800       : 6        |****************************************|

     feishin       : count     distribution
        400        : 2        |********************                    |
        3400       : 4        |****************************************|

     mpv           : count     distribution
        400        : 1        |********************                    |
        4400       : 2        |****************************************|

     sway          : count     distribution
        4600       : 2        |****************************************|

     foot          : count     distribution
        400        : 1        |****************************************|

     Socket Thread : count     distribution
        400        : 1        |****************************************|


     syswide       : count     distribution
        400        : 4151     |****************************************|
        3400       : 607      |*****                                   |
        3800       : 3642     |***********************************     |
        4400       : 607      |*****                                   |
        4600       : 607      |*****                                   |
```

it's so cute :3

**Comment by nekopsykose** [comment]
if you know how to check that the luajit patch 'actually' works, i'll send it to openresty. or you could do it

**Comment by nekopsykose** [comment]
https://github.com/iovisor/bcc/issues/1621#issuecomment-2002537501 yeah it's just broken currently cause that bcc-lua-static variable expands to nothing so none of that is linked. cbf to fix

**Comment by nekopsykose** [comment]
all done then

**Comment by q66** [comment]
the 6.1 kernel builds ok too fwiw

**Comment by q66** [comment]
> yep, it was that easy. but i can't test at runtime since i have no root for bpf on ppc.
> 
> though it doesn't really work anyway:
> 
> ```
> 
> $ doas bcc-lua x.lua
> [ERROR] bcc.lua:2070: Symbol not found: bpf_module_create_c_from_string
> stack traceback:
> 	[C]: in function '__index'
> ```
> 
> those are symbols from bcc source files themselves, so i'm not sure how they're meant to work. in the .lua libbcc comes in via require() and i guess squishy is meant to do some magic. it's not strip/lto that breaks it..

it's probably vis on libbpf that breaks it (hides the symbols) if i had to guess

**Comment by nekopsykose** [comment]
> it's probably vis on libbpf that breaks it (hides the symbols) if i had to guess 

nah, the issue is bcc-lua links bcc source .o's into it for these symbols (they are from bcc, not libbpf), and since they changed it from the .a library to the cmake variable they use for the object list, it doesn't actually work (it's between unrelated cmakelists.txt in different folders, so bcc-lua-static is just an empty list when passed to the bcc-lua target). this looks like it's been broken for years given when that was changed..

**Comment by nekopsykose** [comment]
if you like hacky fixes.. here you go :)

**Comment by nekopsykose** [comment]
i think even if the .o list was not empty the actual linker optimisations would remove all of them anyway as there are no references. i have no idea how this worked for anyone.

anyway, it works now, you can `doas bcc-lua examples/lua/task_switch.lua`` in the repo

**Comment by nekopsykose** [comment]
but tbh i'd rather leave lua disabled

**Comment by q66** [comment]
i'll leave that decision to you

**Comment by nekopsykose** [comment]
yeet

**Comment by q66** [comment]
sure

**Pull request merged**